From 244248f23d93e0f5c66530b5115167f71c9549f8 Mon Sep 17 00:00:00 2001
From: Bart Van Assche <bvanassche@acm.org>
Date: Tue, 19 May 2026 14:21:28 -0700
Subject: [PATCH 174/223] FROMGIT: scsi: ufs: core: Complain if UIC argument 2
 is invalid

According to the UFSHCI standard, the lowest byte of UIC argument 2 is
an output value. Additionally, ufshcd_uic_cmd_compl() is based on the
assumption that the lowest byte of UIC argument 2 is zero. Hence,
complain if the result byte is set when a UIC command is submitted.

Reviewed-by: Peter Wang <peter.wang@mediatek.com>
Signed-off-by: Bart Van Assche <bvanassche@acm.org>
Link: https://patch.msgid.link/20260519212135.3130556-3-bvanassche@acm.org
Signed-off-by: Martin K. Petersen <martin.petersen@oracle.com>
(cherry picked from commit 9fb4c793223b618da6bee50840746522a7da226e https://git.kernel.org/pub/scm/linux/kernel/git/mkp/scsi.git for-next)
Bug: 503802344
Bug: 517489432
Change-Id: I1570102827ecc029e98236a46b4dfa61eec4f025
Signed-off-by: Bart Van Assche <bvanassche@google.com>
(cherry picked from commit 78449bbba352f5afa6d2f825868df41b06eeefcc)
---
 drivers/ufs/core/ufshcd.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/ufs/core/ufshcd.c b/drivers/ufs/core/ufshcd.c
index cb59ae5d5516..776bd33f69db 100644
--- a/drivers/ufs/core/ufshcd.c
+++ b/drivers/ufs/core/ufshcd.c
@@ -2545,6 +2545,7 @@ ufshcd_dispatch_uic_cmd(struct ufs_hba *hba, struct uic_command *uic_cmd)
 	lockdep_assert_held(&hba->uic_cmd_mutex);
 
 	WARN_ON(hba->active_uic_cmd);
+	WARN_ON_ONCE(uic_cmd->argument2 & MASK_UIC_COMMAND_RESULT);
 
 	hba->active_uic_cmd = uic_cmd;
 
-- 
2.50.1

